### PR TITLE
[JetBrains] Show Workspace CPU/Memory resources in control center

### LINKED ISSUE #10580: Integrate with JetBrains backend control center
JetBrains backend control center shows information about the node, not the container, this is due to the control center not respecting applications using cgroups (see: [CWM-6090](https://youtrack.jetbrains.com/issue/CWM-6090/Respect-cgroups-in-performance-indicator)). There should be API's that we can use to add custom performance information for Gitpod into this control center so that it's more usable for users in a Gitpod context. 

**Note:** It would be best if JetBrains are able to update the integration to respect cgroups, however due to time limitations we might want to implement ourselves (Gitpod) and eventually update if JetBrains do get around to updating the performance center to respect cgroups and container-based implementations. 

See: 
* https://www.jetbrains.com/help/idea/remote-development-a.html#control_center

image

Relates to: 
- https://github.com/gitpod-io/gitpod/issues/11042

### PR DESCRIPTION
**Pull request by andreafalzetti**
## Context
Currently, in JetBrains IDE the CPU/Memory usage displayed applies to the whole node, instead of displaying the workspace's specific resources. 

## Description
With this PR we introduce two additional progress bars at the top displaying:

- `Workspace CPU`
- `Workspace Memory`

**Please note**: The Unit used for Memory resource is on purpose "GB" instead of "Mi" like in `gitpod-cli`, my reasons:
1. personally I find it shorter and easier to read `0.38GB / 1.55GB` instead of `2944Mi/11870Mi`
2. when talking about memory (in general) it's more common to refer to GB
3. in gitpod.io our memory (especially for XL) can be a large number to read if it's in `Mi`

I would actually like to propose a separate issue to change it also in `gp top` but I would like to know if there are any reasons or string objections in favor of the smaller unit. cc @akosyakov @felladrin 

## Related Issue(s)
Fixes #10580

## How to test
1. Signup to [this prev env](https://afalz-10581fa3bcc8cc.preview.gitpod-dev.com/workspaces)
5. Select `IntelliJ` + `Latest Release (Unstable)` as your preferred IDE
6. Start a workspace for https://github.com/gitpod-io/spring-petclinic
7. Click on the control center and see available resources, notice the additional rows at the top (Workspace CPU/Memory) before the default CPU/Memory

_OPTIONAL:_

9. To stress the CPU, run: `yes > /dev/null &` multiple times, observe the progress bar growing and moving to Warning and Danger.
10. To get the CPU back to a lower value, run: `killall yes`
11. To stress the Memory, run `cat /dev/zero | head -c 5G | tail` once or twice, and observe the memory progress bar behaving as expected

Screenshot 2022-08-11 at 21 42 39

Screenshot 2022-08-11 at 21 51 25

## Release Notes

```release-note
JetBrains: Display Workspace CPU/Memory usage in Backend Control Center
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

**Comment by roboquat** [comment]
@andreafalzetti: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/ide/jetbrains/backend-plugin/src/main/resources-latest/META-INF/extensions.xml [hunk 1]
```diff
@@ -7,5 +7,7 @@
     <extensions defaultExtensionNs="com.intellij">
         <projectService serviceImplementation="io.gitpod.jetbrains.remote.latest.GitpodTerminalService" client="guest" preload="true"/>
         <projectService serviceImplementation="io.gitpod.jetbrains.remote.latest.GitpodPortForwardingService" preload="true"/>
+        <gateway.customization.performance implementation="io.gitpod.jetbrains.remote.latest.GitpodMetricControlProvider"/>
```
Comment (akosyakov) [inline components/ide/jetbrains/backend-plugin/src/main/resources-latest/META-INF/extensions.xml:10]: i think it would make sense to put them before standard controllers, there should be an attribute to specify before in plugin.xml

File: components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/latest/GitpodMetricControlProvider.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+           createCpuControl(this, backendDiagnosticsService, lifetime)
+            createMemoryControl(this, backendDiagnosticsService, lifetime)
+        }
+    }
+
+    private fun createCpuControl(ctx: VerticalGridBuilder, backendDiagnosticsService: BackendDiagnosticsService, lifetime: Lifetime) {
+        val data = createMetricProgressWithLabels(backendDiagnosticsService, lifetime, "gitpod_workspace_cpu_total", "gitpod_workspace_cpu_used", "gitpod_workspace_cpu_percentage")
+
+        val label = "Workspace CPU"
+        // TODO: make status dynamic based on a threshold
```
Comment (akosyakov) [inline components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/latest/GitpodMetricControlProvider.kt:34]: We need to create an issue to provide severity from supervisor api directly to make sure each client see it the same way. But it should be another PR.

For now we can use 0.8 for warning and 0.95 for danger. I know that in gp cli we have lower values, but i think they are too low there.

File: components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/latest/GitpodMetricControlProvider.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    override fun getControl(lifetime: Lifetime): BeControl {
+        return verticalGrid {
+            val backendDiagnosticsService = BackendDiagnosticsService.Companion.getInstance()
+           createCpuControl(this, backendDiagnosticsService, lifetime)
+            createMemoryControl(this, backendDiagnosticsService, lifetime)
+        }
+    }
+
+    private fun createCpuControl(ctx: VerticalGridBuilder, backendDiagnosticsService: BackendDiagnosticsService, lifetime: Lifetime) {
+        val data = createMetricProgressWithLabels(backendDiagnosticsService, lifetime, "gitpod_workspace_cpu_total", "gitpod_workspace_cpu_used", "gitpod_workspace_cpu_percentage")
```
Comment (akosyakov) [inline components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/latest/GitpodMetricControlProvider.kt:31]: nice that we shod total and use for CPU as well, it would help to clarify for users how much CPU they actually have

File: components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt [hunk 4]
```diff
@@ -365,4 +365,31 @@ class GitpodManager : Disposable {
...
+            try {
+                val f = status.resourcesStatus(Status.ResourcesStatuRequest.getDefaultInstance())
+                resourceStatus = f.asDeferred().await()
+            } catch (t: Throwable) {
+                if (t is CancellationException) {
+                    throw t
+                }
+                thisLogger().error("gitpod: failed to retrieve resource status: ", t)
+            }
+            delay(1000L)
```
Comment (andreafalzetti) [inline components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt:386]: I've noticed the rate-limiting sometimes preventing to receive the fresh metrics. In the meantime that we work on supervisor for extending the rate-limiting capabilities, shall we reduce the refresh to `5s`?

Comment (akosyakov) [reply components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt:386]: It would be implement exponential backoff, you can check how gitpod-monitor does it: https://github.com/akosyakov/gitpod-monitor/blob/15bc5a39d8ab4f71a18ce7e0a38a70c8e9e0e526/src/extension.ts#L65-L77

Comment (akosyakov) [reply components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt:386]: @andreafalzetti Could you create an issue please that we improve stability of this supervisor endpoint? I think we should move polling each second in supervisor and returned cached value to all clients. It will greatly simplify clients.

**Review by akosyakov — APPROVED** [review]
code wise it looks good, but i did not try, i wanted but it was pretty slow, probably someone else already trying :D 

/hold

**Comment by akosyakov** [comment]
> I would actually like to propose a separate issue to change it also in gp top but I would like to know if there are any reasons or string objections in favor of the smaller unit. cc @akosyakov @felladrin

I think it should be aligned with default of standard tools like `top` or `htop`.

File: components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt (hunk 4 above)
Comment (andreafalzetti) [reply components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt:386]: Issue created, I would appreciate a double check of the description to make sure it makes sense 😄 [[1]](https://github.com/gitpod-io/gitpod/issues/12083)

**Comment by andreafalzetti** [comment]
> I think it should be aligned with default of standard tools like top or htop.

It's an interesting take, to look at `htop` or `top` - I had a look as I didn't remember what unit the use, for memory they also use `GB`

Screenshot 2022-08-12 at 10 34 54

**Comment by felladrin** [comment]
> I would actually like to propose a separate issue to change it also in gp top but I would like to know if there are any reasons or string objections in favor of the smaller unit. cc @akosyakov @felladrin

For me, it's also nicer to have it in "GB".

**Comment by felladrin** [comment]
ℹ️ I'm reviewing it now.

**Review by felladrin — APPROVED** [review]
Working great! ✅ 

image

Screenshot 2022-08-12 at 11 56 46

**Comment by andreafalzetti** [comment]
Thanks @felladrin and @akosyakov 🙌 

/unhold

**Comment by akosyakov** [comment]
> For me, it's also nicer to have it in "GB".

let's go with `GB` then
